MOM6: +Expand dimensional consistency testing

Hallberg-NOAA opened the pull request:
This PR includes a series of changes that extends the dimensional consistency
testing, particularly of variables related to density.  Some bugs have been
noted with comments but not corrected yet.  All answers are bitwise identical,
and there are no changes to MOM_parameter_doc or other output files. MOM6
commits with this PR include:

- NOAA-GFDL/MOM6@629416b +Rescaled the units of tv%salt_deficit
- NOAA-GFDL/MOM6@d518449 Corrected units in MOM_hor_visc code comments
- NOAA-GFDL/MOM6@0d43bc3 Corrected units in parameterization code comments
- NOAA-GFDL/MOM6@8ade6de +Pass MLDdensityDifference in units of [R]
- NOAA-GFDL/MOM6@f1b9c66 +Pass timestep to btstep in units of [T]
- NOAA-GFDL/MOM6@59b18f0 +Pass timestep to btstep in units of [T]
- NOAA-GFDL/MOM6@55a0484 Fixed a rescaling factor in entrainment_diffusive
- NOAA-GFDL/MOM6@e55ad23 +Pass timestep to step_forward_MEKE in units of [T]
- NOAA-GFDL/MOM6@5312d99 Cleaned up code in insert_brine
- NOAA-GFDL/MOM6@a4da592 +Rescaled the units of fluxes%heat_content_... vars
- NOAA-GFDL/MOM6@d5ee19a Store timestep in [T} in sum_output_CS
- NOAA-GFDL/MOM6@ecf5813 +Pass timestep to thickness_diffuse in units of [T]
- NOAA-GFDL/MOM6@3f69350 +Pass timestep to extractFluxes1d in units of [T]
- NOAA-GFDL/MOM6@5484ab0 +Pass timestep to insert_brine in units of [T]
- NOAA-GFDL/MOM6@35b884d Fixed a recently added bug in insert_brine
- NOAA-GFDL/MOM6@1d3e21e Merge branch 'dev/gfdl' into density_rescale
- NOAA-GFDL/MOM6@838bc41 Removed unused code in MOM_barotropic.F90
- NOAA-GFDL/MOM6@3b6ccd8 Rescaled RHO0 to [R] in four modules
- NOAA-GFDL/MOM6@8f3c126 Rescaled FLUXCONST to [Z T-1] in eight modules
- NOAA-GFDL/MOM6@cbe3996 Refactored solo_driver/MOM_surface_forcing.F90
- NOAA-GFDL/MOM6@be8e18c +Rescaled the units of surface salt fluxes
- NOAA-GFDL/MOM6@37d5405 Combined scaling factors in MOM_forcing_type
- NOAA-GFDL/MOM6@965c2c2 +Rescaled _rate variables from extractFluxes1d
- NOAA-GFDL/MOM6@76172d4 Simplified scaling factors in MOM_forcing_type
- NOAA-GFDL/MOM6@dbad998 +Rescaled the units of surface mass fluxes
- NOAA-GFDL/MOM6@34193e9 +Added scale arguments in MOM_spatial_means
- NOAA-GFDL/MOM6@fa35255 +Rescaled the units of taux_shelf and tauy_shelf
- NOAA-GFDL/MOM6@3b4ea9c +Turned heat budget pointers into arrays
- NOAA-GFDL/MOM6@1175786 Rescaled density units in MOM_regridding.F90
- NOAA-GFDL/MOM6@e9b36cc +Rescaled density units in coord_slight.F90
- NOAA-GFDL/MOM6@d7f08f5 +Rescaled density units in coord_hycom.F90
- NOAA-GFDL/MOM6@dfd5278 +Changed units of fluxes%TKE_tidal to [R Z3 T-3]
- NOAA-GFDL/MOM6@fb47d7b +Changed set_derived_forcing_fields Rho0 arg units
- NOAA-GFDL/MOM6@79d004b Rescaled gustiness in MOM_surface_forcing files
- NOAA-GFDL/MOM6@1cbf498 +Rescaled units of forces%taux to [R Z L T-2]
- NOAA-GFDL/MOM6@a669294 +Rescaled density units in coord_rho.F90
- NOAA-GFDL/MOM6@e2e4bcb +Rescaled taux_bot and tauy_bot
- NOAA-GFDL/MOM6@08980a7 Merge branch 'dev/gfdl' into density_rescale
- NOAA-GFDL/MOM6@15a5cb5 Rescaled density units in adjustment_init.F90
- NOAA-GFDL/MOM6@0a8e67f Rescaled density units in benchmark_init.F90
- NOAA-GFDL/MOM6@fbed637 Rescaled density units in Rossby_front_2d_init.F90
- NOAA-GFDL/MOM6@a14bca0 Rescaled density units in ISOMIP_initialization.F90
- NOAA-GFDL/MOM6@e137451 Rescaled density units in DOME_initialization.F90
- NOAA-GFDL/MOM6@45de704 Rescaled density units in MOM_tracer_hor_diff.F90
- NOAA-GFDL/MOM6@d072db2 Rescaled density units in wave_structure
- NOAA-GFDL/MOM6@8efd5ae Rescaled density units in MOM_wave_speed.F90
- NOAA-GFDL/MOM6@594300e Rescaled density units in calc_diagnostic_fields
- NOAA-GFDL/MOM6@27a9edd Rescaled density units in adjust_ssh_for_p_atm
- NOAA-GFDL/MOM6@fa02d85 Rescaled density units in MOM_PressureForce_AFV
- NOAA-GFDL/MOM6@7a65682 +Rescaled density units in MOM_PressureForce_Mont
- NOAA-GFDL/MOM6@34f2c39 Rescaled density units in find_eta
- NOAA-GFDL/MOM6@e4fea0a (*)Simplified convert_thickness with no eqn of state
- NOAA-GFDL/MOM6@235d3b6 Rescaled density units in calc_isoneutral_slopes
- NOAA-GFDL/MOM6@f03495f +Rescaled density units in MOM_T_S_init_from_Z
- NOAA-GFDL/MOM6@36fef34 Rescaled density units in initialize_temp_salt_fit
- NOAA-GFDL/MOM6@58cb2d4 Removed a commented out statement
- NOAA-GFDL/MOM6@401ea2b Rescaled density units in MOM_MEKE.F90
- NOAA-GFDL/MOM6@1865f4f (*)+Rescaled MEKE source element units in MEKE_type
- NOAA-GFDL/MOM6@ef1080b +Rescaled energy units in MOM_internal_tides
- NOAA-GFDL/MOM6@3567989 +Rescaled units of set_int_tide_input variables
- NOAA-GFDL/MOM6@04901cd Rescaled density units in MOM_mixed_layer_restrat
- NOAA-GFDL/MOM6@3fb51fd Rescaled units of FrictWork in MOM_hor_visc.F90
- NOAA-GFDL/MOM6@2e3f9c4 Rescaled density units in MOM_thickness_diffuse
- NOAA-GFDL/MOM6@e5ca06e Rescaled units of dRhodT in applyBoundaryFluxesInOut
- NOAA-GFDL/MOM6@8e231df +Pass mixed layer densities to sponges in [R]
- NOAA-GFDL/MOM6@08231ca Used GV%H_to_RZ to simplify rescalings
- NOAA-GFDL/MOM6@2d1f647 +Rescaled specific volume units of dSV_dT args
- NOAA-GFDL/MOM6@5122e33 +Rescaled density units of cTKE arguments
- NOAA-GFDL/MOM6@68e322b Rescaled density units in MOM_energetic_PBL
- NOAA-GFDL/MOM6@80b2d99 +Partially rescaled the units of TKE_itidal_input
- NOAA-GFDL/MOM6@727199a Corrected dimensions in comments
- NOAA-GFDL/MOM6@4808e3a +Rescaled density units in MOM_geothermal
- NOAA-GFDL/MOM6@87a2c4d Rescaled density units in MOM_tidal_mixing
- NOAA-GFDL/MOM6@3fe1c7b Rescaled density units in diagnoseMLDbyDensityDiff
- NOAA-GFDL/MOM6@af57dab Rescaled density units in MOM_internal_tide_input
- NOAA-GFDL/MOM6@7f8b55b Rescaled density units in MOM_kappa_shear
- NOAA-GFDL/MOM6@ac329db Rescaled density units in MOM_set_diffusivity
- NOAA-GFDL/MOM6@297ffe5 Rescaled density units in MOM_set_viscosity
- NOAA-GFDL/MOM6@f7b4b77 Rescaled density units in MOM_regularize_layers
- NOAA-GFDL/MOM6@ee6baaa +Changed units of GV%Rho0 to [R]
- NOAA-GFDL/MOM6@bc378e3 +Move rescaling of Rlay into set_coord routines
- NOAA-GFDL/MOM6@57794ad +Changed units of GV%Rlay to [R]
- NOAA-GFDL/MOM6@4156851 Rescaled density units in MOM_entrain_diffusive
- NOAA-GFDL/MOM6@535950c Merge branch 'dev/gfdl' into density_rescale
- NOAA-GFDL/MOM6@d4c2dfb +Rescale BML densities via calculate_density calls
- NOAA-GFDL/MOM6@74cc9bb +Add optional scale argument to calculate_density
- NOAA-GFDL/MOM6@8eaa01c Rescaled density units in MOM_bulk_mixedlayer
- NOAA-GFDL/MOM6@5274403 +Add H_to_RZ and RZ_to_H to the verticalGrid_type
- NOAA-GFDL/MOM6@3704a66 +Add density rescaling factors in unit_scale_type
- NOAA-GFDL/MOM6@1016be3 +Pass timestep arguments to vertvisc in [T]

Hallberg-NOAA commented:
This PR is being tested with https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/9124.

marshallward commented:
Did you and @adcroft come to a decision about `dt` vs `dt_in_T`?

marshallward commented:
Sounds like we'll address it in a later PR.  I'll send my fixes over to your branch.

Hallberg-NOAA commented:
This updated PR addresses some concerns from @marshallward and @adcroft about unnecessarily renaming variables (especially dt to dt_in_T).  There should now be fewer code differences with the previous code (but some more differences in comments), while also having slightly more complete dimensional rescaling of the timesteps.

Hallberg-NOAA commented:
This PR is being tested (again) with https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/9358 after correcting some diagnostic rescaling and merging in the latest version of dev/gfdl.

adcroft commented on config_src/mct_driver/mom_surface_forcing_mct.F90:
```diff
@@ -237,9 +237,11 @@ subroutine convert_IOB_to_fluxes(IOB, fluxes, index_bounds, Time, G, US, CS, &
   logical :: restore_sst      !< local copy of the argument restore_temp, if it
                               !! is present, or false (no restoring) otherwise.
-  real :: delta_sss           !< temporary storage for sss diff from restoring value
+  real :: delta_sss           !< temporary storage for sss diff from restoring value2
```
Was this "2" intentional? Just curious...

Hallberg-NOAA replied:
No, that "2" at the end of a comment was a mistake. Sorry about that.  Although I tried to be as careful as possible, it would be helpful if someone from NCAR (@gustavo-marques ?) would please evaluate whether the few necessary changes to the mct_driver and nuopc_driver code are as benign as they were intended to be.

marshallward replied:
@Hallberg-NOAA Do you want to wait for feedback from MCT/NUOPC devs before merging?

marshallward commented:
There's a parameter file fail here, does this need a `dev/gfdl` merge?

marshallward commented:
Scratch my previous comment, but noting here that this does make a parameter change (`R_RESCALE_POWER`) and needs a MOM6-examples update.
